src/v/rpc/transport.cc

```diff
@@ -142,38 +145,74 @@ transport::make_response_handler(
+    b.set_correlation_id(idx);
+
+    auto entry = std::make_unique<response_entry>();
+
+    // set initial timing info for this request
+    entry->timing.timeout = opts.timeout;
+    entry->timing.enqueued_at = timing_info::clock_type::now();
+
+    auto handler_raw_ptr = &entry->handler;
     // capture the future _before_ inserting promise in the map
     // in case there is a concurrent error w/ the connection and it
     // fails the future before we return from this function
-    auto response_future = item_raw_ptr->get_future();
-    b.set_correlation_id(idx);
-    auto [_, success] = _correlations.emplace(idx, std::move(item));
+    auto response_future = handler_raw_ptr->get_future();
+
+    auto [_, success] = _correlations.emplace(idx, std::move(entry));
     if (unlikely(!success)) {
         throw std::logic_error(
           fmt::format("Tried to reuse correlation id: {}", idx));
     }
-    item_raw_ptr->with_timeout(opts.timeout, [this, idx, seq] {
-        /*
-         * remove pending entry from requests queue. If a timeout occurred
-         * before an entry was sent we can not keep the entry alive as it may
-         * contain caller semaphore units, the units must be released when we
-         * notify caller with the result.
-         */
-        _requests_queue.erase(seq);
-        auto it = _correlations.find(idx);
-        if (likely(it != _correlations.end())) {
-            vlog(
-              rpclog.info,
-              "Request timeout to {}, correlation id: {} ({} in flight)",
-              server_address(),
-              idx,
-              _correlations.size());
-            _probe.request_timeout();
-            _correlations.erase(it);
-        }
-    });
+    handler_raw_ptr->with_timeout(
+      opts.timeout, [this, method = b.name(), idx, seq] {
+          auto format_ms = [](clock_type::duration d) {
+              auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
+                d);
+              return fmt::format("{} ms", ms.count());
+          };
+
+          auto from_now =
+            [now = timing_info::clock_type::now(), format_ms](
+              timing_info::clock_type::time_point earlier) -> std::string {
+              if (earlier == timing_info::unset) {
+                  return "unset";
+              }
+              return format_ms(now - earlier);
+          };
+          /*
+           * remove pending entry from requests queue. If a timeout occurred
+           * before an entry was sent we can not keep the entry alive as it may
+           * contain caller semaphore units, the units must be released when we
+           * notify caller with the result.
+           */
+          _requests_queue.erase(seq);
+          auto it = _correlations.find(idx);
+          // The timeout may race with the completion of the request (and
+          // removal from _correlations map) in which case we treat this as a
+          // not-timed-out request.
```

Comment:
this comment is referring to the else case for the immediately following conditional?

Reply:
Yes, exactly. Could I make this clearer? Something like:

```
// The timeout has fired and we expect to find the request still in the correlation
// map (the true case below). However, the timeout does race with the completion
// of the request (at which point it is removed from _correlations map), so if don't
// find the correlation ID we simply assume this was a successful request (or at 
// least one which was removed from the map for some other reason).